object-api-multipart.go

```diff
@@ -106,51 +107,52 @@ func (o objectAPI) listMetaVolumeFiles(prefixPath string, markerPath string, rec
 		// Loop through and validate individual file.
 		for _, fi := range fileInfos {
 			var entries []FileInfo
-			// List all the entries if fi.Name is a leaf directory, if
-			// fi.Name is not a leaf directory them the resulting
-			// entries are empty.
-			entries, e = o.listLeafEntries(fi.Name)
-			if e != nil {
-				log.WithFields(logrus.Fields{
-					"prefixPath": fi.Name,
-				}).Errorf("%s", e)
-				return nil, false, e
+			if fi.Mode.IsDir() {
+				// List all the entries if fi.Name is a leaf directory, if
+				// fi.Name is not a leaf directory them the resulting
+				// entries are empty.
+				entries, e = o.listLeafEntries(fi.Name)
+				if e != nil {
+					log.WithFields(logrus.Fields{
+						"prefixPath": fi.Name,
+					}).Errorf("%s", e)
+					return nil, false, e
+				}
 			}
 			// Set markerPath for next batch of listing.
 			markerPath = fi.Name
 			if len(entries) > 0 {
+
+				// We reach here for non-recursive case and a leaf entry.
+
 				for _, entry := range entries {
-					// Skip the entries for erasure parts if any.
-					if strings.Contains(path.Base(entry.Name), ".") {
-						continue
-					}
 					allFileInfos = append(allFileInfos, entry)
+					newMaxKeys++
+					if newMaxKeys == maxKeys {
+						// eof value is automatically taken care of.
+						return
+					}
 				}
-			} else {
-				// Skip special files.
+				continue
+			}
+
+			// We reach here for a non-recursive case non-leaf entry
+			// OR recursive case with fi.Name matching pattern bucket/object/uploadID[.partNum.md5sum]
+
+			if !fi.Mode.IsDir() { // Do not skip non-recursive case directory entries.
+				// Skip files matching pattern bucket/object/uploadID.partNum.md5sum
+				// and retain files matching pattern bucket/object/uploadID
 				specialFile := path.Base(fi.Name)
 				if strings.Contains(specialFile, ".") {
 					// Contains partnumber and md5sum info, skip this.
 					continue
 				}
-				allFileInfos = append(allFileInfos, fi)
 			}
+			allFileInfos = append(allFileInfos, fi)
 			newMaxKeys++
-			// If we have reached the maxKeys, it means we have listed
-			// everything that was requested. Return right here.
 			if newMaxKeys == maxKeys {
-				// Returns all the entries until maxKeys entries.
-				//
```

Comment:
Keep the context of the comment, remove the cruft about eof being set deliberately. "eof value is automatically taken care of" has no meaning.